### ISSUE TITLE
[Feature] Add annotations to Backup Job template

### ISSUE THREAD

**Issue by andreasgeisslerdt** [mariadb-operator/mariadb-operator#350]
**Is your feature request related to a problem? Please describe.**
I am using mariadb-operator in a ServiceMesh environment (Istio), which adds to each pod a envoy proxy sidecar.
Setting some general annotations/labels the mariadb pods are well running.

In case of the Backup a job is created. 
In the job pod also a sidecar is injected, which is not deleted after the backup container finished, which leads to the situation, that the job will never be finished. (https://medium.com/redbox-techblog/handling-istio-sidecars-in-kubernetes-jobs-c392661c4af7)
To overcome this problem you have 2 possibilities:
1. add a sidecar container to kill the envoy proxy after the job container finishes
2. add a label to the job template to disable the sidecar injection
```
  jobTemplate:
    spec:
      template:
        metadata:
          labels:
            sidecar.istio.io/inject: "false"
```
The option 2 would help in case of mariadb-operator.

**Describe the solution you'd like**

I suggest to add an option in the Backup CRD (https://github.com/mariadb-operator/mariadb-operator/blob/main/docs/API_REFERENCE.md#backupspec) to define additional annotations for job templates

**Describe alternatives you've considered**
The alternative was to force the istio platform to disable the sidecar injection for the specific jobs.
But this is not working, as there is no fixed jobname:
```
sidecarInjectorWebhook:
  # You can use the field called alwaysInjectSelector and neverInjectSelector which will always inject the sidecar or
  # always skip the injection on pods that match that label selector, regardless of the global policy.
  # See https://istio.io/docs/setup/kubernetes/additional-setup/sidecar-injection/#more-control-adding-exceptions
  neverInjectSelector:
    - matchExpressions:
        - {key: job-name, operator: In, values: ["..."]}
```

**Environment details**:
- Kubernetes version: 1.27.5
- Kubernetes distribution: Vanilla
- mariadb-operator version: 0,24,0
- Install method: helm via ArgoCD
- Install flavor: recommended

**Additional context**

**Comment by mmontes11** [comment]
Hey there @andreasgeisslerdt ! Thanks for reporting this.

I agree, the option 2 seems better to integrate with `mariadb-operator`. You can actually achive this by specifying `spec.inheritMetadata` in the `MariaDB` resource:

```yaml
apiVersion: mariadb.mmontes.io/v1alpha1
kind: MariaDB
metadata:
  name: mariadb
spec:
  # metadata that will be inherited by all children objects
  inheritMetadata:
    labels:
      database.myorg.io: mariadb  
    annotations:
      database.myorg.io: mariadb

  volumeClaimTemplate:
    annotations:
      database.myorg.io: mariadb
    resources:
      requests:
        storage: 1Gi
    accessModes:
      - ReadWriteOnce
```

and then, when you create a `Backup` referring to it:

```yaml
apiVersion: mariadb.mmontes.io/v1alpha1
kind: Backup
metadata:
  name: backup
spec:
  mariaDbRef:
    name: mariadb
  storage:
    persistentVolumeClaim:
      resources:
        requests:
          storage: 100Mi
      accessModes:
        - ReadWriteOnce
  args:
    - --single-transaction
    - --all-databases
```

The following `Pod` will eventually be created:

```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    database.myorg.io: mariadb
  creationTimestamp: "2024-01-25T16:52:32Z"
  finalizers:
  - batch.kubernetes.io/job-tracking
  generateName: backup-
  labels:
    batch.kubernetes.io/controller-uid: 28688b01-21bd-4cb7-8ce1-dd1fe8599259
    batch.kubernetes.io/job-name: backup
    controller-uid: 28688b01-21bd-4cb7-8ce1-dd1fe8599259
    database.myorg.io: mariadb
    job-name: backup
  name: backup-2sc28
...
``` 

Which I believe it should cover this case, could you confirm please? 🙏🏻

**Comment by andreasgeisslerdt** [comment]
Hi Martin,
the suggestion does not cover it completely.
When I set the label to `spec.inheritMetadata`, it will affect all resources, which means, that the sidecars are disabled for all pods and Istio is disabled for the whole component.
This is actually what I am currently doing, as otherwise the Metrics export is not working.
A better way I actually saw in other operators (get for Postgres) where I struggled with the same problem.
https://access.crunchydata.com/documentation/postgres-operator/latest/references/crd/5.5.x/postgrescluster
You can see, that in each level (PostgresCluster, PostgresCluster.spec,  PostgresCluster.spec.backups,...) you have an "metadata" entry to add annotations and labels to the resources, which works really fine.
I just added for the PostgresCluster.spec.backups my lable to disable the sidecar injection and the created jobs were deployed without sidecars.
I know, that's maybe a big change, but could be an improvement idea for the mariadb-operator.
Thanks a lot for your support...
Actually I would like to help you in implementation, but I am currently too busy with my job
Best regards
Andreas

**Comment by mmontes11** [comment]
Hey there @andreasgeisslerdt ! 

Understood, I see your point, I've created the following issue for this:
- https://github.com/mariadb-operator/mariadb-operator/issues/353

> Actually I would like to help you in implementation, but I am currently too busy with my job

Contributions are always welcome! I will leave some time till the end of the release to see if someone decides to contribute and otherwise will try to squeeze it in my tasks for this release.

**Comment by mmontes11** [comment]
Hey there @addreas ! 

This has been addressed. You can now define dedicated metadata for `Backup`, `Restore` and `SqlJob` objects via their new `inheritMetadata` field. The metadata is propagated to the `Pod` template as well. Please take a look at the updated examples:

- https://github.com/mariadb-operator/mariadb-operator/blob/main/examples/manifests/mariadb_v1alpha1_backup.yaml

Closing! This will be released in `v0.0.26`. Stay tuned!
